feat(core): add ability to resolve mount targets using EFS API

jusiskin opened the pull request:
Resolves #391

Taken from that issue:

> The RFDK code to mount EFS file-systems on an instance assumes that the instance belongs to a VPC with Amazon-supplied DNS (see [VPC DNS documentation](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-dns.html)). When a VPC is configured to disable the Amazon-supplied DNS resolver, then the DNS names used by both the EFS mount helper and the NFS mounting code cannot be resolved and the mount will fail with an error showing up in the `cloud-init` log file, such as:

> ```
> [   32.413572] cloud-init[3739]: Failed to resolve "fs-a1b2c3d4.efs.us-west-2.amazonaws.com" - check that your file system ID is correct.
> ```

## Solution

The DNS names in the error message above correspond to [EFS mount  targets](https://docs.aws.amazon.com/efs/latest/ug/manage-fs-access.html). These provide the endpoints that serve the networked file-system traffic. EFS automatically adds DNS entries into your VPC's Amazon-supplied Route 53 resolver, but when this feature is disabled, those names cannot be resolved using DNS.

Fortunately, the IP address of each mount target can be obtained by calling the EFS [`DescribeMountTargets` API](https://docs.aws.amazon.com/efs/latest/ug/API_DescribeMountTargets.html).

Added an optional boolean flag when EFS mounting script that will perform this lookup and bake the IP address into the host machine's `/etc/hosts` file to bypass DNS lookups. Using this looks like:

**TypeScript:**

```ts
new MountableEfs(scope, {
  // ...
  resolveMountTargetDnsWithApi: true,
});
```

**Python:**

```py
MountableEfs(scope,
  # ...
  resolve_mount_target_dns_with_api=True,
)
```

## Testing

Modified the `All-In-AWS-Infrastructure-Basic` example, enabling the above flag. Connected to the mounting instances using SSM Session Manager and confirmed:

*   The `/etc/hosts` file contains an entry for the mount target:

        10.0.71.79 fs-a1b2c1d2.efs.us-west-2.amazonaws.com # Added by RFDK

*   The EFS file-system driver establishes a connection to the specified IP address:

    ```sh
    $ sudo netstat -alnp | grep 10.0.71.79
    tcp        0      0 10.0.0.13:55102         10.0.71.79:2049         ESTABLISHED 2306/stunnel
    ```

*   Deployment succeeds
*   Am able to connect to the farm, submit a job, view successful render logs (requires proper EFS mounting on the `RenderQueue`)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

jusiskin committed:
feat(core): add ability to resolve mount targets using EFS API

ddneilson commented on packages/aws-rfdk/lib/core/lib/mountable-efs.ts:
```diff
@@ -66,6 +66,18 @@ export interface MountableEfsProps {
    */
   readonly extraMountOptions?: string[];
+
+  /**
```
There's a downside here that we should mention... If you have mount targets in multiple AZs, then the Route53 resolver will always give you the mount target IP that's in the same AZ as the requester (or random if the request isn't coming from an IP that shares an AZ with a resolution).

When resolving via IP address like this you don't get that AZ-matching. Thus, cross-AZ data charges may be incurred.

jusiskin replied:
This code actually does performs AZ matching. See:

https://github.com/jusiskin/aws-rfdk/blob/daa2fd714859be3cf31c45a394ec883f8dfa7f94/packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh#L78

and

https://github.com/jusiskin/aws-rfdk/blob/daa2fd714859be3cf31c45a394ec883f8dfa7f94/packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh#L85

ddneilson replied:
Yeah. I saw that, but didn't come back to remove the comment. My bad.

ddneilson commented on packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh:
```diff
@@ -58,6 +63,68 @@ function use_nfs_mount() {
+if [[ $RESOLVE_MOUNTPOINT_IP_VIA_API == "true"  ]]
+then
+  # jq is used to query the JSON API response
+  sudo "${PACKAGE_MANAGER}" install -y jq
```
Should check whether `jq` is present, and only hit the package manager if it is not. Use-case is an isolated machine with no network connectivity to the package repository.

jusiskin replied:
Good call! I'll add that check first

ddneilson commented on packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh:
```diff
@@ -58,6 +63,68 @@ function use_nfs_mount() {
+
+  if [[ $MNT_TARGET_RESOURCE_ID == fs-* ]]; then
+    # Mounting without an access point
+    MOUNT_POINT_IP=$(aws efs describe-mount-targets \
```
Missing failure handling -- we're not guaranteed that there is a MountTarget in the same AZ as the instance. Route53 would resolve the DNS with a randomly chosen IP in that case.

Same in the `fsap-*` case

jusiskin replied:
Given the concern you mentioned about cross-AZ data charges, do you think we should replicate this behavior? Or is it better to fail? I'm learning towards failing with a user-friendly warning over using a random mount target. That way anyone using this could avoid the cross-AZ costs and simply add the missing mount target.

The calling code has a catch-all error handler, but you're right - currently this will not fail gracefully if there's no matching mount target. I'll fix that - thanks!

ddneilson replied:
My opinion... We want to make sure that the customer knows that there could be cross-AZ data charges, but not prevent them from deploying in a cross-AZ manner. Throwing an error for cross-AZ would make a cross-AZ mount impossible, and there may be valid use-cases for that.

jusiskin replied:
I've added logic to fall-back to using any available mount target with a warning being output that should get logged to CloudWatch. Let me know what you think.

ddneilson commented on packages/aws-rfdk/lib/core/lib/mountable-efs.ts:
```diff
@@ -157,11 +169,26 @@ export class MountableEfs implements IMountableLinuxFilesystem {
       'pushd "$TMPDIR"',
       `unzip ${mountScript}`,
-      `bash ./mountEfs.sh ${this.props.filesystem.fileSystemId} ${mountDir} ${mountOptionsStr}`,
+      `bash ./mountEfs.sh ${this.props.filesystem.fileSystemId} ${mountDir} ${mountOptionsStr} ${resolveMountTargetDnsWithApi}`,
```
I suggest reversing the ordering of mount options & the new arg. Reason being that mount options could be empty, but the new arg will never be.

jusiskin replied:
done

ddneilson commented on packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh:
```diff
@@ -13,12 +13,15 @@
 if test $# -lt 2
 then
-  echo "Usage: $0 <file system ID> <mount path> [<mount options>]"
+  echo "Usage: $0 FILE_SYSTEM_ID MOUNT_PATH [MOUNT_OPTIONS] [RESOLVE_MOUNT_POINT_USING_API]"
```
Makes sense to me to make the new arg required. Two optional positional args can lead to confusion.

jusiskin replied:
applied

jusiskin committed:
chore(core): apply MR feedback

- Make RESOLVE_MOUNT_POINT_USING_API argument required and move before optional mount opts
- Add graceful error-handling when no matching mount target found for instance's availability zone

jusiskin committed:
chore(core): remove sudo from mountEfs.sh

- userdata runs as root and some AMIs might not have sudo installed

horsmand commented on packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh:
```diff
@@ -49,31 +55,104 @@ else
+
+  # Get access point ID if available, otherwise file system ID
+  MNT_TARGET_RESOURCE_ID=$FILESYSTEM_ID
+  ACCESS_POINT_MOUNT_OPT=$(echo "${MOUNT_OPTIONS}" | sed -e 's#,#\n#g' | grep 'accesspoint=') || true
```
We're creating a dependency on the format of the mount options as we define them in the MountableEfs construct here and I could see that format getting changed without testing this specific use case. Maybe we should just add a comment where we define those mount options mentioning this?

jusiskin replied:
This is parsing a standard options string as passed to the mount system call (see `-o` option in the [docs for the CLI command equivalent](https://man7.org/linux/man-pages/man8/mount.8.html)).

Discussion offline, the confusion was around the `accesspoint=...` format, which is what the [EFS mount helper expects](https://docs.aws.amazon.com/efs/latest/ug/efs-mount-helper.html#mounting-access-points) (see docs). Added a comment here with a link to those docs.

jusiskin committed:
chore(core): fallback to using any mount target with a warning

ddneilson commented on packages/aws-rfdk/lib/core/scripts/bash/mountEfs.sh:
```diff
@@ -12,13 +12,17 @@
 
 set -xeu
 
 if test $# -lt 2
```
```suggestion
if test $# -lt 3
```

jusiskin replied:
good catch - fixed

ddneilson reviewed (changes requested):
One small nit, and then I think we're good.

jusiskin committed:
chore(core): fix validation for number of required args in mountEfs.sh

Co-authored-by: Daniel Neilson <53624638+ddneilson@users.noreply.github.com>
